src/main/java/org/craftercms/engine/util/spring/security/preview/ConfigAwarePreviewAccessTokenFilter.java

```diff
@@ -113,9 +120,63 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
             throw new PreviewAccessException(HttpStatus.FORBIDDEN, message);
         }
 
+		// Create preview token and site name cookies to support preview workflow
+		int maxAge = Math.max((int)((tokenExpiryTimestamp - System.currentTimeMillis()) / 1000), 0);
+		if (tokenFromQueryParam && maxAge > 0) {
+			createPreviewCookie(httpServletRequest, httpServletResponse, previewToken, maxAge);
+			createSiteNameCookie(httpServletRequest, httpServletResponse, maxAge);
+		}
+
         chain.doFilter(request, response);
     }
 
+	/**
+	 * Creates a preview cookie with the given token.
+	 *
+	 * @param request the HTTP request
+	 * @param response the HTTP response
+	 * @param previewToken the preview token
+	 * @param  maxAge the max-age value
+	 */
+	private void createPreviewCookie(HttpServletRequest request, HttpServletResponse response, String previewToken, int maxAge) {
+		createCookie(request, response, PREVIEW_SITE_TOKEN_NAME, previewToken, maxAge);
+	}
+
+	/**
+	 * Creates a cookie with the site name.
+	 *
+	 * @param request the HTTP request
+	 * @param response the HTTP response
+	 * @param maxAge the max-age value
+	 */
+	private void createSiteNameCookie(HttpServletRequest request, HttpServletResponse response, int maxAge) {
+		String siteName = request.getParameter(siteNameParam);
+		if (isEmpty(siteName)) {
+			return;
+		}
+
+		createCookie(request, response, siteNameParam, siteName, maxAge);
+	}
+
+	/**
+	 * Creates a cookie with the given name and value.
+	 *
+	 * @param request the HTTP request
+	 * @param response the HTTP response
+	 * @param name the name of the cookie
+	 * @param value the value of the cookie
+	 * @param maxAge the max-age value
+	 */
+	private void createCookie(HttpServletRequest request, HttpServletResponse response, String name, String value, int maxAge) {
```

Comment:
I think Path, HttpOnly and SameSite should be configurable

Reply:
Updated